Add routing strategy and fallback configuration support for gateway endpoints

TomeHirata opened the pull request:
## 🥞 Stacked PR
Use this [link](https://github.com/mlflow/mlflow/pull/19483/files) to review incremental changes.
- [**stack/gateway/fallback**](https://github.com/mlflow/mlflow/pull/19483) [[Files changed](https://github.com/mlflow/mlflow/pull/19483/files)]

---------
<!-- Remove unused checkboxes except for the patch release section at the bottom -->

### Related Issues/PRs

n/a

### What changes are proposed in this pull request?

This PR introduces traffic routing and fallback to the gateway endpoints. We will have two types of configuration:

1. Traffic routing strategy: how to pick the first destination from the primary model definitions. Currently only weight based traffic split is supported.
2. Fallback strategy: how the request is propagated when the request for the first destination fails. Currently sequential model fallback is supported.



### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [x] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- Do not modify or remove any text inside the parentheses. Keep both checkboxes below. -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

TomeHirata committed:
Add routing strategy and fallback configuration support for gateway endpoints

This commit introduces a new routing strategy for gateway endpoints, allowing for fallback configurations. The `RoutingStrategy` and `FallbackStrategy` enums have been added, along with the `FallbackConfig` message to define fallback behavior. The `GatewayEndpoint` entity has been updated to include these new fields, and the necessary changes have been made across various components, including the database models, API handlers, and providers. Additionally, tests have been added to ensure the correct functionality of the fallback provider.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
type hint

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
migration

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
test

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
test

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
http status

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
schema

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

BenWilson2 commented on mlflow/entities/gateway_endpoint.py:
```diff
@@ -22,6 +25,77 @@ class GatewayResourceType(str, Enum):
+
+    Args:
+        strategy: The fallback strategy to use (e.g., FallbackStrategy.SEQUENTIAL).
+        max_attempts: Maximum number of models to try.
```
Is the logic for fallback "try primary one time, if fail, go to next" or is it "try primary n times, if fail, got to next for n times try"?

TomeHirata committed:
Add linkage type and fallback order to gateway endpoint model mappings

This commit introduces a new `LinkageType` enum to differentiate between PRIMARY and FALLBACK linkages in the `GatewayEndpointModelMapping`. It also adds a `fallback_order` field to specify the order of fallback attempts. The changes include updates to the database schema, API handlers, and various components to support these new fields, ensuring proper handling of fallback configurations in gateway endpoints.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
lint

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

B-Step62 commented on mlflow/entities/gateway_endpoint.py:
```diff
@@ -22,6 +25,77 @@ class GatewayResourceType(str, Enum):
+        return ProtoRoutingStrategy.Value(self.value)
+
+
+class FallbackStrategy(str, Enum):
```
Yea it's fine, the confusion mainly came from the fact that routing strategy had contained fallback.

TomeHirata committed:
test

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
inventory

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
test

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

B-Step62 commented on mlflow/entities/gateway_endpoint.py:
```diff
@@ -22,6 +25,77 @@ class GatewayResourceType(str, Enum):
+        try:
+            return cls(ProtoRoutingStrategy.Name(proto))
+        except ValueError:
+            return None
```
Got it, can we add some comment for the context?

B-Step62 commented on mlflow/server/gateway_api.py:
```diff
@@ -157,12 +164,203 @@ def _create_provider_from_endpoint_name(
+            error_code=RESOURCE_DOES_NOT_EXIST,
+        )
+
+    model_definition_ids = fallback_config.model_definition_ids
```
Is it guaranteed that the `model_definition_ids` is sorted by the fallback order? If not, I think we need to sort it here.

B-Step62 reviewed (approved):
LGTM overall, left a few minor comments and suggestions

TomeHirata committed:
comments

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
update fallback logic

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
Refactor gateway endpoint fallback configuration

This update modifies the fallback configuration for gateway endpoints by replacing the `model_definition_ids` field with `model_mappings`, which now holds a list of `GatewayEndpointModelMapping` objects. The changes also include updates to the related protobuf definitions, SQLAlchemy models, and REST API methods to accommodate the new structure. Additionally, tests have been updated to reflect these changes, ensuring proper functionality of the fallback logic.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
handler test

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
fix

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
use structured config

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
test

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
Refactor conditional checks for fallback and model definition IDs

This update simplifies the conditional checks in the `SqlAlchemyGatewayStoreMixin` class by removing explicit `None` checks for `fallback_model_definition_ids` and `model_definition_ids`. Additionally, test assertions have been updated to reflect changes in the expected number of model mappings for gateway endpoints, ensuring consistency across the tests.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>
